mmcv/ops/csrc/pytorch/contour_expand.cpp

```diff
@@ -0,0 +1,107 @@
+// It is modified from https://github.com/whai362/PSENet
+#include <iostream>
+#include <queue>
+
+#include "pytorch_cpp_helper.hpp"
+
+using namespace std;
+
+class Point2d {
+ public:
+  int x;
+  int y;
+
+  Point2d() : x(0), y(0) {}
+  Point2d(int _x, int _y) : x(_x), y(_y) {}
+};
+
+void growing_text_line(const uint8_t *data, IntArrayRef data_shape,
+                       const int *label_map,
+
+                       IntArrayRef label_shape, int &label_num, int &min_area,
+                       vector<vector<int>> &text_line) {
+  std::vector<int> area(label_num + 1);
+  for (int x = 0; x < label_shape[0]; ++x) {
+    for (int y = 0; y < label_shape[1]; ++y) {
+      int label = label_map[x * label_shape[1] + y];
+      if (label == 0) continue;
+      area[label] += 1;
+    }
+  }
+
+  queue<Point2d> queue, next_queue;
+  for (int x = 0; x < label_shape[0]; ++x) {
+    vector<int> row(label_shape[1]);
+    for (int y = 0; y < label_shape[1]; ++y) {
+      int label = label_map[x * label_shape[1] + y];
+      if (label == 0) continue;
+      if (area[label] < min_area) continue;
+
+      Point2d point(x, y);
+      queue.push(point);
+      row[y] = label;
+    }
+    text_line.emplace_back(row);
+  }
+
+  int dx[] = {-1, 1, 0, 0};
+  int dy[] = {0, 0, -1, 1};
+
+  for (int kernel_id = data_shape[0] - 2; kernel_id >= 0; --kernel_id) {
```

Comment:
Why the last element is ignored?

Reply:
data_shape[0] is the number of kernel.
data_shape[0]-1 is the minimum internal kernel.
data_shape[0]-2 is the second  smallest internal kernel,  in which we'd like to expand contour at the beginning.

Reply:
Okay, it is the smallest kernel, and need nothing to do